_posts/2020-02-09-git-guide.markdown

```diff
@@ -0,0 +1,162 @@
+## 合并多个提交commit
+### 需求分析：
+1. 发现了很多typo，分了好几次才全部修改完成，自己看不下去
+2. 热爱提交，commit琐碎重复，被老大要求合并成一个commit
+3. 根据网上教程进行了同步或合并，结果不仅merge了别人的commit，还没有合并成功自己的commit
+### git合并多个提交命令
+- 查看log    
+`git log`  
+![gitlog](https://img-blog.csdnimg.cn/2020020917373270.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L1NpbmR3ZWxsZXI1NTMw,size_16,color_FFFFFF,t_70)  
+这里会出现很多个commit，找到你需要合并的commit，比如将最新一个（HEAD所指）commit和第二个合并为一次提交，复制作为父亲的提交后面那一串 c47e54a3c35b6cec9f5fd2be0fc3905785df6df0之类的    
+- 重置提交状态到父提交时，之后会在这个基础上提交一次新的、合并了所有需要合并的提交的提交  
+`git reset xxxxx`  
+- 重新推到暂存区  
+`git add -A`  
+- 对提交内容描述进行修改  
+`git commit --amend`    
+会进入文本编辑，修改描述，之后的commit就只有这次描述了，合并的commit里的旧描述可以注释掉    
+![amend](https://img-blog.csdnimg.cn/20200209173705731.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L1NpbmR3ZWxsZXI1NTMw,size_16,color_FFFFFF,t_70)  
+- push到origin  
+`git push -f origin xxx`  
+`-f`的作用是强制推送到orign，因为前面在本地合并commit之后，对应的远程分支中的commit并没有变化，git会认为远程分支的提交领先于本地分支  
+![-f](https://img-blog.csdnimg.cn/20200210124806165.jpg)  
+强制推送之后即可合并远程分支上的commit  
+刷新github看看commit记录  
+## 代码版本回退、合并
+### 需求分析：
+1. 开发到一半，upstream突然更新了
+2. 开发前忘记更新本地分支
+### git回退与合并命令
+- 先拉取upstream  
+`git fetch upstream`  
+- 需要版本回退，用upstream/master更新本地代码  
+`git rebase upstream/master`  
+- 或者需要合并本地代码和上游代码  
+`git merge upstream/master`  
+> 可能会遇到各种冲突，根据git提示来就行，直接打开vscode也会给你标出冲突的地方，选择就行  
+> `git rebase`比`git merge`要好，想要更好的提交树，使用`rebase`操作会更好一点。这样可以线性的看到每一次提交，并且没有增加提交节点。  
+> `merge`操作遇到冲突的时候，当前`merge`就不能继续进行了。手动修改冲突内容后，`add`提交修改，再`commit`就可以了。  
+> 而`rebase`操作的话，会中断`rebase`,同时会提示去解决冲突。解决冲突后,将修改`add`后执行`git rebase –continue`继续操作，或者`git rebase –skip`忽略冲突。 
+## 修改pr
+直接修改对应的本地分支再push到origin对应分支就行了，pr会自动跟着branch的commit变
+## 总结
+综上，一套完整的开发、提交pr流程一般为：  
+1. 新建本地目录，后续操作都需要在这个目录下进行
+2. 第一种情况：  
+> 如果不存在远程仓库，即为还未上传的本地项目，在该目录下，`git init`初始化本地git仓库，这个命令用一次就行了，之后基于该仓库的操作请勿再使用init命令，人生没有重来，本地仓库也是
+> `git remote add origin ssh_url` ssh_url是复制自己的仓库的ssh,将本地仓库关联该远程仓库
+> `git remote -v`查看关联的仓库  
+
+第二种情况：  
+
+> Gituhub上已有项目远程仓库
+> `git clone ssh_url`将代码下载下来，建议使用fork的origin仓库  
+> `git remote add upstream ssh_url`关联上游仓库
+> `git remote -v`查看关联的仓库  
+> `git fetch upstream`+`git rebase upstream/master`开发之前一定要确保本地代码与上游同步，是最新的  
+3. `git checkout -b feature-A`基于local/master，新建本地分支，开发功能A
+4. 开发ing……
+5. 开发完成后，`git status`看看修改的文件是不是那几个，确认后`git add -A`把所有的文件（代码）加入暂存区（或者`git add [<path>]`将指定文件或目录加入暂存区）
+6. `git commit -m "the changes"`添加一次提交，""内填写本次提交概述
+7. `git push origin branch_name`将刚才加入暂存区的文件（代码）和提交内容推到origin的branch_name分支中
+8. 在github网页上，将origin/branch_name分支以pr形式提交给上游仓库
+9. 如若需要进行下一个功能的开发，记得切换回master分支，再新建本地feature-B分支进行B功能的开发
+### 注意！
+为国争光，禁用pull，从我做起。
+> `git pull`是一个非常强硬的命令，它会忽略一切细节并且可以把`git fetch`、`git merge`和`git rebase`等等的活儿都干了，中间没有出于安全考虑的确认  
+> 该命令还会将合并结果记录在新提交中，包含两个父提交的名称以及来自用户的描述更改的日志消息
+> 虽然简单，但请勿使用`git pull`
+## 流程链接
+[见贡献流程](https://github.com/imjoey/openkylin/wiki)
```

Comment:
Since this repository is private, so this link is unreachable by public. We should use [Collaborating with issues and pull requests](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests) instead.